### PR TITLE
Spec improvements

### LINKED ISSUE #2634: spec: globbing problems
The `spec` plugin doesn't do globbing correctly: 

- The `#` glob should query the array size and should copy metadata to non-existent keys within the array bounds as well. Otherwise something like `kdb get array/#0` won't work, even if the specification contains a default value for `array/#`. Setting the `fallback` metadata might also be a solution.
- The same problem exists for `_`, however, I don't really see a solution there, because you can't create all possible keys.

Where possible, the `spec` plugin should also use the `globbing` library.

### LINKED ISSUE #2700: spec: does not remove copied metadata before writing config
I think this is a show-stopper for using in LCDproc or machinekit (who want INI as first step for smooth migration)

## Steps to Reproduce the Problem

- mount a specification
- mount an INI file
- do a `kdb set` 

```sh
kdb mount wrong-copy.ini /tests ini
kdb setmeta /tests/some-key some-meta some-value
kdb set /tests/some-key x
```

## Expected Result

To have a file

```ini
some-key = x
```

## Actual Result

```ini
#@META some-meta = some-value
some-key = x
```

## System Information

- Elektra Version: master

## Implementation Hints

The spec plugin does not have the positioning to remove its meta-data before writing to the storage. See discussion below: the KeySet passed back from `kdbSet` **must not** be altered. Only the meta-data within `split` is allowed to be removed.

So the easiest way is to call the spec plugin for every part in the split plugin separately. (As foreach plugin.)

### PR DESCRIPTION
**Pull request by kodebach**
**NOTE:** Technically there are some breaking changes in this PR. However since, every breaking change that I know of concerns parts of `spec` that can already be considered broken, I don't think that is a problem.

---

Closes #2634 (only array part, new issue for _ needed)

Also improves `ksIsBelow` and friends (see #2727) and `ksSearchInternal` (see #2455). Changing `ksAppend` is more complicated, because we need to make sure the OPMPHM is not invalidated, when we append a subset of the current KeySet (i.e. no new keys).

`elektraUnescapeKeyName` now only does the full escaping procedure, if needed and `ksCut` utilises `ksSearchInternal` to find the start point of the cut.

## Basics

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [x] Longer descriptions should be in documentation or in design decisions.
- [x] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

- [x] I added unit tests
- [x] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see doc/CODING.md)
- [x] meta data is updated (e.g. README.md of plugins and doc/METADATA.md)

## Review

@markus2330 please review my pull request

### CODE REVIEW

**Comment by markus2330** [comment]
This pull request **introduces 4 alerts** when merging a4437ed459052d348509f1198aed2c0cd7fd6e25 into 5f7f855c79753f734946ef1a964570cba311e189 - [view on LGTM.com](https://lgtm.com/projects/g/ElektraInitiative/libelektra/rev/pr-f9e7e4689a2c91ceb9a0913a704020ee6d92f8c7)

**new alerts:**

* 4 for Array argument size mismatch

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by kodebach** [comment]
I decided not to fix #2700 in this PR, because that needs more considerations (see https://github.com/ElektraInitiative/libelektra/issues/2700#issuecomment-497977597).

File: doc/AUTHORS.md [hunk 1]
```diff
@@ -195,3 +195,11 @@ development of the mmapstorage plugin
 - email: mpranj@limun.org
 - github user: [mpranj](https://github.com/mpranj)
 - devel/test on: Fedora, Debian, macOS
+
+## Klemens Böswirth
+
+development of the highlevel API and code-generation; various other things
```
Comment (markus2330) [inline doc/AUTHORS.md]: Good that you finally add yourself! :heartbeat:

File: src/libs/elektra/internal.c [hunk 2]
```diff
@@ -715,6 +715,21 @@ size_t elektraUnescapeKeyName (const char * source, char * dest)
 
 	ELEKTRA_ASSERT (sp != NULL && dp != NULL, "Got null pointer sp: %p dp: %p", (void *) sp, (void *) dp);
 
+	// if there is nothing to unescape, just make a copy and replace / with \0
+	if (strpbrk (sp, "\\%") == NULL)
```
Comment (markus2330) [inline src/libs/elektra/internal.c]: Is this a performance improvement? Why not put it into a separate function?

File: src/plugins/spec/README.md [hunk 3]
```diff
@@ -1,76 +1,138 @@
...
+### Array Size Validation
+
+As hinted to above, we validate array sizes. If a spec key `x/#` is given, and the spec key `x` has the metakey `array/min` or `array/max`
+set, we validate the array size (given as metakey `array` on `x`) is within the limits of `array/min` and `array/max`. Both `array/min` and
+`array/max` have to be valid array-elements similar to `array`. If the array size is out of bounds, this causes a `range` conflict.
+
+We also check that the array only contains actual array elements. If this not the case, this causes a `member` conflict.
 
-### Per Key Configuration
+Note: We don't actually validate that the array doesn't contain elements above the given array size. This is because it doesn't have anything
```
Comment (markus2330) [inline src/plugins/spec/README.md]: Good that you wrote the note, this is unexpected.

File: src/libs/elektra/internal.c (hunk 2 above)
Comment (kodebach) [reply src/libs/elektra/internal.c]: Yes, this should improve performance. The function was so short already, it didn't seem necessary to split it up.

**Pull request merged**